src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -433,50 +448,63 @@ public synchronized void unsubscribe(UnsubscribeRequest request)
-                publishRequests.add(getConnection(false).publish(m, request.getQos(), request.isRetain())
-                                .whenComplete((packetId, throwable) -> {
+        MqttMessage m = new MqttMessage(request.getTopic(), request.getPayload());
+        return getConnection(false).publish(m, request.getQos(), request.isRetain())
+                .whenComplete((packetId, throwable) -> {
                     // packetId is the SDK assigned ID. Ignore this and instead use the spooler ID
                     if (throwable == null) {
                         spool.removeMessageById(id);
                         logger.atDebug().kv("id", id).kv("topic", request.getTopic())
-                            .log("Successfully published message");
+                                .log("Successfully published message");
                     } else {
                         spool.addId(id);
                         logger.atError().log("Failed to publish the message via Spooler", throwable);
                     }
-                }));
+                });
+    }
 
-                if (publishRequests.size() >= maxInFlightPublishes) {
-                    CompletableFuture.anyOf(
-                            publishRequests.toArray(new CompletableFuture[0])).get();
-                    publishRequests = publishRequests.stream().filter(f -> !f.isDone()).collect(Collectors.toList());
+    /**
+     * Iterate the spooler queue to publish all the spooled message.
+     */
+    @SuppressWarnings("PMD.AvoidCatchingThrowable")
+    protected void publishAllTasks() {
+        List<CompletableFuture<?>> publishRequests = new ArrayList<>();
+        while (!Thread.currentThread().isInterrupted()) {
+            try {
+                getConnection(false).connect().get();
+                while (mqttOnline.get()) {
+                    CompletableFuture future = publishSingleTask();
+                    if (future == null) {
+                        continue;
+                    }
+                    publishRequests.add(future);
+
+                    if (publishRequests.size() >= maxInFlightPublishes) {
+                        CompletableFuture.anyOf(
+                                publishRequests.toArray(new CompletableFuture[0])).get();
+                        publishRequests = publishRequests.stream().filter(f ->
+                                !f.isDone()).collect(Collectors.toList());
+                    }
+                }
+                break;
+            } catch (ExecutionException e) {
+                logger.atError().log("Error when publishing from spooler", e);
+                if (Objects.equals(Utils.getUltimateCause(e).getClass(), InterruptedException.class)) {
+                    Thread.currentThread().interrupt();
+                    logger.atDebug().log("Shutting down spooler task");
                 }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
```

Comment:
You don't need to interrupt here. You just need to exit the loop which will make the thread stop since it has nothing to do.

Reply:
Exception for the InterrupedException, it would not stop. IF the queue is empty, it will wait until there is new message.

Reply:
There's more than one way to exit a loop.

Reply:
Instead of interrupting the thread, just break out of the loop.